### ISSUE TITLE
Incorrect `on_classical_vals` overrides for non classical bloqs

### ISSUE THREAD

**Issue by anurudhp** [quantumlib/Qualtran#1515]
I ran into this while implementing #1514. The tensors of these bloq examples seem to have non-trivial phases, which makes them non-classical. The following bloq examples fail in #1514 (I've restricted to bloqs with up to 9 qubits):
```
sub_from_small

E Mismatched elements: 144 / 65536 (0.22%)
E Max absolute difference: 1.
E Max relative difference: 1.
E x: array([[1., 0., 0., ..., 0., 0., 0.],
E [0., 1., 0., ..., 0., 0., 0.],
E [0., 0., 1., ..., 0., 0., 0.],...
E y: array([[1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
E [0.+0.j, 1.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
E [0.+0.j, 0.+0.j, 1.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],...
```
```
cmod_add_k_small

E Mismatched elements: 18 / 1024 (1.76%)
E Max absolute difference: 1.
E Max relative difference: 0.
E x: array([[1., 0., 0., ..., 0., 0., 0.],
E [0., 1., 0., ..., 0., 0., 0.],
E [0., 0., 1., ..., 0., 0., 0.],...
E y: array([[1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
E [0.+0.j, 1.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
E [0.+0.j, 0.+0.j, 1.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],...
```
```
moddbl_small

E Mismatched elements: 3 / 256 (1.17%)
E Max absolute difference: 1.
E Max relative difference: 1.
E x: array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
E [0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
E [0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],...
E y: array([[1.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j,
E 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
E [0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 1.+0.j,...
```
```
approx_cswap_small

E Mismatched elements: 120 / 262144 (0.0458%)
E Max absolute difference: 2.
E Max relative difference: 2.
E x: array([[1., 0., 0., ..., 0., 0., 0.],
E [0., 1., 0., ..., 0., 0., 0.],
E [0., 0., 1., ..., 0., 0., 0.],...
E y: array([[1.000000e+00+6.471125e-32j, 0.000000e+00+0.000000e+00j,
E 0.000000e+00+0.000000e+00j, ..., 0.000000e+00+0.000000e+00j,
E 0.000000e+00+0.000000e+00j, 0.000000e+00+0.000000e+00j],...
```

**Comment by mpharrigan** [comment]
Do you know where the non-trivial phases are being introduced? Presumably if you write a decomposition of e.g. ModDbl using only classical operations (whose matrices only have `+1` entries) then you can't get any phases popping up, right?

**Comment by tanujkhattar** [comment]
fyi -- `approx_cswap_small`'s tensor would have a non-trivial phases since it implements an approximate cswap which is correct upto a phase, in order to have 4 T count instead of 7,  which gets inverted when the approx cswap inverse is applied. This is similar to the 4 T version of Toffoli which applies a toffoli upto a relative phase (we don't explicitly have a bloq for this yet)

**Comment by mpharrigan** [comment]
Yeah, the approx cswap is clearer. It's got "approx" in the name, so it's less surprising that there are phases. The other ones are surprising to me

**Comment by NoureldinYosri** [comment]
`sub_from_small ` => the classical action of `SubtractFrom` is wrong. it uses `fmod` which gives wrong answer on overflow instead it should call `simulation.classical_sim.add_ints` ... this also means that the classical action tests for this bloq are insufficient.

cmod_add_k_small and moddbl_small are because their constructions assume correct range ... that they assume they will be called on states $x < \mod < 2^\mathrm{bitsize}$ ... outside this range (i.e. $\mod \leq x < 2\mathrm{bitsize}) which happens for tensor contraction their constructions doen't work .... I can change the `on_classical_vals` to take this into account (if input is outside the valid range return it as is) ... this will make the tensor contraction work but that may introduce a difference in behaviour between the bloq and it decomposition for those out of bound inputs

**Comment by mpharrigan** [comment]
is it obvious how those corner cases lead to _phases_ in the tensors?

**Comment by anurudhp** [comment]
I looked at the failing tests again, and I realized that the arithmetic bloqs fail because the inputs outside range (i.e. `>= mod`) do not match (the max abs diff is `1`, which is from `1`s and `0`s mismatching).

Only the `cswap_approx` fails because of a phase mismatch.

**Comment by NoureldinYosri** [comment]
> is it obvious how those corner cases lead to _phases_ in the tensors?

there are no phases except in `cswap_approx`. as anurudhp this is mismatch is between 0 and 1 for inputs outside the valid range for the bloqs

**Comment by mpharrigan** [comment]
ok, that makes way more sense
